FAPI: Support intermediate certificates stored in NV ram.

JuergenReppSIT opened the pull request:
* Add reading of intermediate certificates from NV ram.
* Add Intel ODCA Root Certificate.
* Rework Fapi_GetPlatformCertificates.
    
Addresses: #2934

ben-krieger commented:
@JuergenReppSIT and @AndreasFuchsTPM 

I tested this out on a 12th gen system (Intel N100), but it segfaulted and produced different error codes.

---

Before patch:

```Dockerfile
FROM ghcr.io/tpm2-software/ubuntu-22.04:latest

RUN git clone --depth=1 https://github.com/tpm2-software/tpm2-tss && \
    cd tpm2-tss && \
    ./bootstrap && ./configure --prefix=/ && make -j && make -j install

RUN git clone --depth=1 https://github.com/tpm2-software/tpm2-tools && \
    cd tpm2-tools && \
    ./bootstrap && ./configure --prefix=/ && make -j && make -j install

ENTRYPOINT /bin/bash
```

```console
$ tpm2_getcap handles-nv-index
- 0x1C00002
- 0x1C0000A
- 0x1C00016
- 0x1C00100
$ tss2_provision
WARNING:fapi:src/tss2-fapi/ifapi_io.c:419:ifapi_io_check_create_dir() Directory /root/.local/share/tpm2-tss/user/keystore does not exist, creating
WARNING:fapi:src/tss2-fapi/ifapi_io.c:419:ifapi_io_check_create_dir() Directory //var/lib/tpm2-tss/system/keystore/policy does not exist, creating
ERROR:fapi:src/tss2-fapi/ifapi_curl.c:222:ifapi_curl_verify_ek_cert() ErrorCode (0x00060025) No certificate downloaded
ERROR:fapi:src/tss2-fapi/api/Fapi_Provision.c:974:Fapi_Provision_Finish() ErrorCode (0x00060025) Verify EK certificate
ERROR:fapi:src/tss2-fapi/api/Fapi_Provision.c:182:Fapi_Provision() ErrorCode (0x00060025) Provision
Fapi_Provision(0x60025) - fapi:No certificate
```

---

After patch:

```Dockerfile
FROM ghcr.io/tpm2-software/ubuntu-22.04:latest

RUN git clone --depth=1 https://github.com/tpm2-software/tpm2-tss && \
    cd tpm2-tss && git fetch origin pull/2943/head:pr2943 && git switch pr2943 && \
    ./bootstrap && ./configure --prefix=/ && make -j && make -j install

RUN git clone --depth=1 https://github.com/tpm2-software/tpm2-tools && \
    cd tpm2-tools && \
    ./bootstrap && ./configure --prefix=/ && make -j && make -j install

ENTRYPOINT /bin/bash
```

```console
$ tss2_provision
WARNING:fapi:src/tss2-fapi/ifapi_io.c:419:ifapi_io_check_create_dir() Directory /root/.local/share/tpm2-tss/user/keystore does not exist, creating
WARNING:fapi:src/tss2-fapi/ifapi_io.c:419:ifapi_io_check_create_dir() Directory //var/lib/tpm2-tss/system/keystore/policy does not exist, creating
ERROR:fapi:src/tss2-fapi/ifapi_verify_cert_chain.c:197:ifapi_verify_cert_chain() EK verification failed
ERROR:fapi:src/tss2-fapi/api/Fapi_Provision.c:997:Fapi_Provision_Finish() Failed to verify certificate chain. ErrorCode (0x00060001)
ERROR:fapi:src/tss2-fapi/api/Fapi_Provision.c:185:Fapi_Provision() ErrorCode (0x00060001) Provision
Fapi_Provision(0x60001) - fapi:Catch all for all errors not otherwise specified
Segmentation fault (core dumped)
```

JuergenReppSIT commented:
@ben-krieger Thank you for testing the draft PR.  It would be very helpful if you could upload the certificates stored at 0x1C20000 that you get with:
`tpm2_nvread 0x1C20000 -o certs`
I tried to simulate the verification error. But I could not reproduce the segmenation fault which seems to occur in the tpm2_tools part of tss2_provision with the current master of the tools. Which version of the tpm2 tools are you using?

ben-krieger commented:
> @ben-krieger Thank you for testing the draft PR. It would be very helpful if you could upload the certificates stored at 0x1C20000 that you get with: `tpm2_nvread 0x1C00100 -o certs` I tried to simulate the verification error. But I could not reproduce the segmenation fault which seems to occur in the tpm2_tools part of tss2_provision with the current master of the tools. Which version of the tpm2 tools are you using?

Here's the requested output. I converted `certs` to PEM encoding.

```text
-----BEGIN CERTIFICATE-----
MIICwDCCAkagAwIBAgIQZh8cb2xgi11PF6iu9nM0ozAKBggqhkjOPQQDAzBIMS4w
LAYDVQQLDCVPRENBIDIgQ1NNRSBQX0FETCAwMDAwMjcwMSBJc3N1aW5nIENBMRYw
FAYDVQQDDA13d3cuaW50ZWwuY29tMB4XDTIxMDkwMTAwMDAwMFoXDTQ5MTIzMTIz
NTk1OVowGjEYMBYGA1UEAwwPQ1NNRSBBREwgUk9NIENBMHYwEAYHKoZIzj0CAQYF
K4EEACIDYgAEYxpAsa2haORWG1Bzbl4e33bcxKgB+bQ5fLZEONqQFUS3wN9cftFl
iCrE0AChGDTxtZJ+8ziVrSHt6sGJlmSW0NxSDNMO0i+x594Ef4ip6qYEzSdIvInX
+EibulYjeha+o4IBITCCAR0wHwYDVR0jBBgwFoAUelQ/9soR0JlnnDamEfJh7xXV
IJIwHQYDVR0OBBYEFHkXldcu7GJrdxy9b2DYXBDSriMGMA8GA1UdEwEB/wQFMAMB
Af8wDgYDVR0PAQH/BAQDAgGuMGIGCCsGAQUFBwEBBFYwVDBSBggrBgEFBQcwAoZG
aHR0cHM6Ly90c2NpLmludGVsLmNvbS9jb250ZW50L09uRGllQ0EvY2VydHMvQURM
XzAwMDAyNzAxX09EQ0FfQ0EyLmNlcjBWBgNVHR8ETzBNMEugSaBHhkVodHRwczov
L3RzY2kuaW50ZWwuY29tL2NvbnRlbnQvT25EaWVDQS9jcmxzL0FETF8wMDAwMjcw
MV9PRENBX0NBMi5jcmwwCgYIKoZIzj0EAwMDaAAwZQIwCrim/1RsBYHJozEK0yfV
OSinDWWaesTOFCZQ2WFvmQi10RWPt6Bc0pFsetJvB+VvAjEAu+38dkiYyemwk3kO
2UJv3CU1xvtJ1bfpKoSRPtjKcNExUJZwaZM5nd1hwP9Cm2bu
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
MIICczCCAfmgAwIBAgIBATAKBggqhkjOPQQDAzAaMRgwFgYDVQQDDA9DU01FIEFE
TCBST00gQ0EwHhcNMjIwODA5MDAwMDAwWhcNNDkxMjMxMjM1OTU5WjAjMSEwHwYD
VQQDDBhDU01FIEFETCBTVk4wMSBLZXJuZWwgQ0EwdjAQBgcqhkjOPQIBBgUrgQQA
IgNiAAQUY6NLkNT1o3Ow/TzHlIZnZU0hZUTAuPtFHI4YuipP3T82oSP5oyN5+ivf
nFKoCIY1SFbQTJrsb99Q2r5n6CxNeSngADBuyeTeqJ5ytWnnsXkeuLZfhALugAUM
zB0DI+ejggEIMIIBBDAfBgNVHSMEGDAWgBR5F5XXLuxia3ccvW9g2FwQ0q4jBjAd
BgNVHQ4EFgQU1bYG/DF1NsYWl+QGUyCsHQ9PaU4wDwYDVR0TAQH/BAUwAwEB/zAO
BgNVHQ8BAf8EBAMCAqwwgaAGA1UdHwSBmDCBlTCBkqBKoEiGRmh0dHBzOi8vdHNj
aS5pbnRlbC5jb20vY29udGVudC9PbkRpZUNBL2NybHMvT0RDQV9DQTJfQ1NNRV9J
bmRpcmVjdC5jcmyiRKRCMEAxJjAkBgNVBAsMHU9EQ0EgQ0EyIENTTUUgSW50ZXJt
ZWRpYXRlIENBMRYwFAYDVQQDDA13d3cuaW50ZWwuY29tMAoGCCqGSM49BAMDA2gA
MGUCMAOQl1lKPyJz0gsXA1fq8pzfAJIPChnkuWW99cehUdqMiXvlxJkTA6boXx8N
O4A0QAIxANH0/4e3NWdmCe0Cqkyt+ha4URd8BQ+AhzIvxbbRqTbWFc7aLIE+XNge
9EpUlR+kfA==
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
MIICejCCAgGgAwIBAgIFAQQTAEAwCgYIKoZIzj0EAwMwIzEhMB8GA1UEAwwYQ1NN
RSBBREwgU1ZOMDEgS2VybmVsIENBMB4XDTIyMDgwOTAwMDAwMFoXDTQ5MTIzMTIz
NTk1OVowHjEcMBoGA1UEAwwTQ1NNRSBBREwgUFRUICAwMVNWTjB2MBAGByqGSM49
AgEGBSuBBAAiA2IABK49LD8p9CspIQJqo6OAC47XN31FUvbVOF8wVzFmtAsDRPd4
NS47l9Bsla/0PKxgcw6CNgXTg222ypdOWhQI/knGob3m3sbq8uV0dX4QZ3X9nZfp
CvPzZ5TofOGXnl9BD6OCAQgwggEEMB8GA1UdIwQYMBaAFNW2BvwxdTbGFpfkBlMg
rB0PT2lOMB0GA1UdDgQWBBT6L6i0TGaYdNVvQ6hrKF4jNygcsjAPBgNVHRMBAf8E
BTADAQH/MA4GA1UdDwEB/wQEAwICBDCBoAYDVR0fBIGYMIGVMIGSoEqgSIZGaHR0
cHM6Ly90c2NpLmludGVsLmNvbS9jb250ZW50L09uRGllQ0EvY3Jscy9PRENBX0NB
Ml9DU01FX0luZGlyZWN0LmNybKJEpEIwQDEmMCQGA1UECwwdT0RDQSBDQTIgQ1NN
RSBJbnRlcm1lZGlhdGUgQ0ExFjAUBgNVBAMMDXd3dy5pbnRlbC5jb20wCgYIKoZI
zj0EAwMDZwAwZAIwUW/dfuNZ9546/onH/8Z0IsJocwWkLsYsoU/QzNCZAwtXkZYH
BLIzw8EweZhzzv+aAjB2BYv587cPuVPjpE6cujvC1qU8BlwqqqCuHR1Y45pCjXZv
MV9Z8E0WNe5SgbqUVrs=
-----END CERTIFICATE-----
```

Based on the `Dockerfiles` shared previously, I'm just using tpm-tools from `master` (which has not changed in 2 weeks).

ben-krieger commented:
> @ben-krieger @AndreasFuchsTPM Thank you for uploading the certificates. I did expect that it would be sufficient to store the OnDie root certificate and the certificates from 0x1C00100 into the ossl store. But there are two more intermediate certificates referenced which have to be downloaded. I will add the download as soon as I find time....

Wow, so the benefits of having the cert chain locally-stored is mostly negated. That's a frustrating design choice. Though if the 2 intermediates that need to be downloaded are fairly stable over time, then it's not terrible.

I guess I'd just make sure that intermediates are only downloaded when necessary.

Thank you for this work!

JuergenReppSIT commented:
@ben-krieger @AndreasFuchsTPM I have added the download of the intermediate certificates. It would be great if you could test it again. If there are still errors it would be interesting to get the output of:
`TSS2_LOG=fapi+debug tss2_provision `
Thank you for the help.

ben-krieger commented:
Still getting the same errors, but without the segfault at the end.

```console
$ tss2_provision
WARNING:fapi:src/tss2-fapi/ifapi_io.c:419:ifapi_io_check_create_dir() Directory /root/.local/share/tpm2-tss/user/keystore does not exist, creating
WARNING:fapi:src/tss2-fapi/ifapi_io.c:419:ifapi_io_check_create_dir() Directory //var/lib/tpm2-tss/system/keystore/policy does not exist, creating
ERROR:fapi:src/tss2-fapi/ifapi_verify_cert_chain.c:347:ifapi_verify_cert_chain() EK verification failed
ERROR:fapi:src/tss2-fapi/api/Fapi_Provision.c:1000:Fapi_Provision_Finish() Failed to verify certificate chain. ErrorCode (0x00060001)
ERROR:fapi:src/tss2-fapi/api/Fapi_Provision.c:185:Fapi_Provision() ErrorCode (0x00060001) Provision
Fapi_Provision(0x60001) - fapi:Catch all for all errors not otherwise specified
```

The full debug logs: [debug.log](https://github.com/user-attachments/files/19212863/debug.log)

ben-krieger commented:
> @ben-krieger Thank you for uploading the debug information. All intermediate certificates are downloaded and the chain is completed with the OnDie root certificate. So the verification should actually work but unfortunately it didn't. Could you please upload the EK certificate stored at 0x01c0000a? (if you use the default profile P_ECCP256SHA256) So I could try to create a test case with the complete chain and also verify the chain with the openssl command line tools.

Here's the EK cert. I immediately notice that the issuer name and SKI match, which is a good sign and I'm not sure where validation fails.

```
-----BEGIN CERTIFICATE-----
MIICsjCCAjigAwIBAgIQc8l48fPt9UKyLNabrAmZIDAKBggqhkjOPQQDAzAeMRww
GgYDVQQDDBNDU01FIEFETCBQVFQgIDAxU1ZOMB4XDTIyMDgwOTAwMDAwMFoXDTQ5
MTIzMTIzNTk1OVowADBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABJcYi7HBfMH9
0M9AMl2uRtGyzf9G2OwJZR3hJKcwxi58hqSL4O6qiRqwRkhtMJuLviYqE2OQygnH
PYn7o3yHSJejggF0MIIBcDAfBgNVHSMEGDAWgBT6L6i0TGaYdNVvQ6hrKF4jNygc
sjAPBgNVHRMBAf8EBTADAQEAMA4GA1UdDwEB/wQEAwIDCDCBoAYDVR0fBIGYMIGV
MIGSoEqgSIZGaHR0cHM6Ly90c2NpLmludGVsLmNvbS9jb250ZW50L09uRGllQ0Ev
Y3Jscy9PRENBX0NBMl9DU01FX0luZGlyZWN0LmNybKJEpEIwQDEmMCQGA1UECwwd
T0RDQSBDQTIgQ1NNRSBJbnRlcm1lZGlhdGUgQ0ExFjAUBgNVBAMMDXd3dy5pbnRl
bC5jb20wEAYDVR0lBAkwBwYFZ4EFCAEwJQYDVR0JAQEABBswGTAXBgVngQUCEDEO
MAwMAzIuMAIBAAICAIowUAYDVR0RAQH/BEYwRKRCMEAxFjAUBgVngQUCAQwLaWQ6
NDk0RTU0NDMxDjAMBgVngQUCAgwDQURMMRYwFAYFZ4EFAgMMC2lkOjAyNTgwMDEy
MAoGCCqGSM49BAMDA2gAMGUCMDuDaIFvQg+iVgJkTJbZgESVvgz9KKvhPQF3yE9x
Fi65aLZ2USZ8Lj+g1V8v7DfL2wIxAMzyWOUz275E0YKjKapWpTqU/qnZe457HHzK
hiR9EssFJNfQ6tnCXj0brqC/70rgKQ==
-----END CERTIFICATE-----
```

JuergenReppSIT commented:
@ben-krieger I could verify the EK certificate with the OnDie root certificate and the five intermediate certificate with a C program which uses the same OSSL functions as FAPI does. So I'm also not sure where the verification fails.
I have improved the debugging output. It would great if you could test it again to produce the improved debugging output. Sorry for inconvenience and thank you for the help.

ben-krieger commented:
> `TSS2_LOG=fapi+debug tss2_provision `

Happy to help.

[debug2.log](https://github.com/user-attachments/files/19233766/debug2.log)

ben-krieger commented:
Seems like maybe it doesn't find:

- `/OU=ODCA CA2 CSME Intermediate CA/CN=www.intel.com`
- `/OU=ODCA 2 CSME P_ADL 00002701 Issuing CA/CN=www.intel.com`

JuergenReppSIT commented:
@ben-krieger Thank you for uploading. 
It looks like this is the output from the old version. With the new version, the EK certificate should be displayed at the end.
Also the download of the two certificates you mentioned should be logged.

ben-krieger commented:
> @ben-krieger Thank you for uploading. It looks like this is the output from the old version. With the new version, the EK certificate should be displayed at the end. Also the download of the two certificates you mentioned should be logged.

Whoops. Try this.
[debug3.log](https://github.com/user-attachments/files/19233995/debug3.log)

ben-krieger commented:
Hmm... `EK verification failed: unable to get certificate CRL`

CRL: https://tsci.intel.com/content/OnDieCA/crls/ODCA_CA2_CSME_Indirect.crl

JuergenReppSIT commented:
In my test C program which simulates the FAPI verification I forgot to activate the CRL check with:
`X509_STORE_set_flags(store, X509_V_FLAG_CRL_CHECK | X509_V_FLAG_CRL_CHECK_ALL);`
Therefore the verification worked in the test but no CRL check was executed. FAPI checks the CRLs. After I added this line to the test program also there I get the error:
`unable to get certificate CRL`
But all CRLs where dowloaded and added to the store:
Downloading CRL from: https://tsci.intel.com/content/OnDieCA/crls/OnDie_CA.crl
Downloading CRL from: https://tsci.intel.com/content/OnDieCA/crls/ODCA_CA2_CSME_Product.crl
Downloading CRL from: https://tsci.intel.com/content/OnDieCA/crls/ADL_00002701_ODCA_CA2.crl
Downloading CRL from: https://tsci.intel.com/content/OnDieCA/crls/ODCA_CA2_CSME_Indirect.crl
Downloading CRL from: https://tsci.intel.com/content/OnDieCA/crls/ODCA_CA2_CSME_Indirect.crl
In tests with IFX TPMs and certificates also CRLs where added to the store and the check worked. That's strange.

JuergenReppSIT commented:
@ben-krieger You mentioned CRL: https://tsci.intel.com/content/OnDieCA/crls/ODCA_CA2_CSME_Indirect.crl
Do you have any idea how to fix this problem?

JuergenReppSIT commented:
@ben-krieger The verification seems to work for the sub chain of certificates which have an "Authority Key Identifier" for which a CRL with the same "Authority Key Identifier"  can be found in the store.

JuergenReppSIT commented:
@ben-krieger I have added the flag X509_V_FLAG_EXTENDED_CRL_SUPPORT to support indirect crls and marked the intermediate certificates as untrusted. But the error still does occur.
Also the verification of the chain with used crls in the command line does not work:
```
openssl verify -CAfile OnDie_CA_RootCA_Certificate.cer.pem \
    -untrusted ODCA_CA2_CSME_Intermediate.cer.pem \
    -untrusted ADL_00002701_ODCA_CA2.cer.pem \
    -untrusted nvram3.cer.pem \
    -untrusted nvram2.cer.pem \
    -untrusted nvram1.cer.pem \
    -CRLfile OnDie_CA.crl.pem \
    -CRLfile ODCA_CA2_CSME_Product.crl.pem \
    -CRLfile ODCA_CA2_CSME_Indirect.crl.pem \
    -CRLfile ADL_00002701_ODCA_CA2.crl.pem \
    -crl_check_all ek.cer.pem
```

JuergenReppSIT commented:
@ben-krieger Good news. It should work now.
For indirect crls it was not sufficient to add the crls to the store. A list of crls has to be added to the store context.
Could you please test it again. I hope it will be the last test.

ben-krieger commented:
@JuergenReppSIT Unfortunately I'm now getting `ERROR:fapi:src/tss2-fapi/ifapi_verify_cert_chain.c:368:ifapi_verify_cert_chain() EK verification failed: unable to get certificate CRL`. But it seems like great progress!

[debug4.log](https://github.com/user-attachments/files/19275689/debug4.log)

JuergenReppSIT commented:
@ben-krieger Thank you very much for your patience during testing. I've now modified the FAPI so that it exactly matches my test program which verifies the EK certificate. I hope it will work now.

ben-krieger commented:
Well, I've run into an unrelated issue: https://github.com/tpm2-software/tpm2-tss/issues/1626

```console
$ tss2_provision
ERROR:fapi:src/tss2-fapi/api/Fapi_Provision.c:1172:Fapi_Provision_Finish() ErrorCode (0x0006000b) SRK persistent handle already defined
WARNING:fapi:src/tss2-fapi/ifapi_io.c:511:ifapi_io_remove_directories() Removing: //var/lib/tpm2-tss/system/keystore/P_ECCP256SHA256/HE/EK/object.json
ERROR:fapi:src/tss2-fapi/api/Fapi_Provision.c:185:Fapi_Provision() ErrorCode (0x0006000b) Provision
Fapi_Provision(0x6000B) - fapi:A parameter has a bad value
```

This is probably a good sign that the EK cert chain part is all fixed now.

AndreasFuchsTPM reviewed (approved):
Seems to work as tested by Intel
